source/common/http/header_map_impl.h

```diff
@@ -230,121 +242,253 @@ class HeaderMapImpl : public virtual HeaderMap, NonCopyable {
+    ASSERT(handle.it_->second < inlineHeadersSize());
+    HeaderEntry& entry = maybeCreateInline(&inlineHeaders()[handle.it_->second], handle.it_->first);
+    subtractSize(entry.value().size());
+    entry.value().setInteger(value);
+    addSize(entry.value().size());
+  }
+  size_t removeInline(Handle handle) override {
+    ASSERT(handle.it_->second < inlineHeadersSize());
+    return HeaderMapImpl::removeInline(&inlineHeaders()[handle.it_->second]);
+  }
+  static size_t inlineHeadersSize() {
+    return StaticLookupTable<Interface>::size() * sizeof(HeaderEntryImpl*);
+  }
+
+protected:
+  absl::optional<StaticLookupResponse> staticLookup(absl::string_view key) override {
+    return StaticLookupTable<Interface>::lookup(*this, key);
+  }
+  virtual const HeaderEntryImpl* const* constInlineHeaders() const PURE;
+};
+
+#define DEFINE_HEADER_HANDLE(name)                                                                 \
+  Handle name =                                                                                    \
+      CustomInlineHeaderRegistry::getInlineHeader<header_map_type>(Headers::get().name).value();
+
+/**
+ * Concrete implementation of RequestHeaderMap which allows for variable custom registered inline
+ * headers.
+ */
+class RequestHeaderMapImpl final : public TypedHeaderMapImpl<RequestHeaderMap>,
+                                   public InlineStorage {
+public:
+  static std::unique_ptr<RequestHeaderMapImpl> create() {
+    return std::unique_ptr<RequestHeaderMapImpl>(new (inlineHeadersSize()) RequestHeaderMapImpl());
+  }
+
   INLINE_REQ_HEADERS(DEFINE_INLINE_HEADER_FUNCS)
   INLINE_REQ_RESP_HEADERS(DEFINE_INLINE_HEADER_FUNCS)
 
-protected:
-  // Explicit inline headers for the request header map.
-  // TODO(mattklein123): This is mostly copied between all of the concrete header map types.
-  // In a future change we can either get rid of O(1) headers completely, or it should be possible
-  // to statically register all O(1) headers and move to a single dynamically sized class where we
-  // we reference the O(1) headers in the table by an offset.
-  struct AllInlineHeaders {
-    AllInlineHeaders() { clear(); }
-    void clear() { memset(this, 0, sizeof(*this)); }
-
-    INLINE_REQ_HEADERS(DEFINE_INLINE_HEADER_STRUCT)
-    INLINE_REQ_RESP_HEADERS(DEFINE_INLINE_HEADER_STRUCT)
+private:
+  struct HeaderHandleValues {
+    INLINE_REQ_HEADERS(DEFINE_HEADER_HANDLE)
+    INLINE_REQ_RESP_HEADERS(DEFINE_HEADER_HANDLE)
   };
 
-  absl::optional<StaticLookupResponse> staticLookup(absl::string_view key) override {
-    return StaticLookupTable<RequestHeaderMapImpl>::lookup(*this, key);
-  }
-  void clearInline() override { inline_headers_.clear(); }
+  using HeaderHandles = ConstSingleton<HeaderHandleValues>;
 
-  AllInlineHeaders inline_headers_;
+  RequestHeaderMapImpl() { clearInline(); }
+  void clearInline() override { memset(inline_headers_, 0, inlineHeadersSize()); }
```

Comment:
The variable length member has to be in the most derived class, so that means that `constInlineHeaders()` and `inlineHeaders()` can't be moved. For the same reason I don't think that we can move `clearInline()` since otherwise we would call a virtual function during construction. (This is a bit of a dark area of C++ so feel free to correct me if there is a better way to do this.)

Reply:
Makes sense. You might be able to factor out the leaf class impls using a template? Definitely could be a follow-up.

In any case adding commentary about the above description would be good.